streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/bolt/query/RealtimeJoinBolt.java

```diff
@@ -0,0 +1,638 @@
+        if(this.lookupStream!=null) {
+            throw  new IllegalArgumentException("Cannot declare a Lookup stream more than once");
+        }
+        this.lookupStream = lookupStream;
+        this.dropOlderDuplicates = dropOlderDuplicates;
+        this.retentionCount = retentionCount.value;
+        this.timeBasedRetention = false;
+        this.joinType = JoinType.OUTER;
+        return this;
+    }
+
+    public RealtimeJoinBolt outerJoin(String lookupStream, BaseWindowedBolt.Duration retentionTime, boolean dropOlderDuplicates) {
+        if(this.lookupStream!=null) {
+            throw  new IllegalArgumentException("Cannot declare a Lookup stream more than once");
+        }
+        if (retentionTime.value<=0) {
+            throw  new IllegalArgumentException("Retention time must be positive number");
+        }
+        this.lookupStream = lookupStream;
+        this.dropOlderDuplicates = dropOlderDuplicates;
+        this.retentionTime = retentionTime.value;
+        this.timeBasedRetention = true;
+        this.joinType = JoinType.OUTER;
+        return this;
+    }
+
+    public RealtimeJoinBolt dataStream(String dataStream) {
+        if (this.dataStream!=null) {
+            throw  new IllegalArgumentException("Cannot declare a Data stream more than once");
+        }
+        this.dataStream = dataStream;
+        return this;
+    }
+
+    /**
+     * Specify output fields
+     *      e.g: .select("lookupField, stream2:dataField, field3")
+     * Nested Key names are supported for nested types:
+     *      e.g: .select("outerKey1.innerKey1, outerKey1.innerKey2, stream3:outerKey2.innerKey3)"
+     * Inner types (non leaf) must be Map<> in order to support nested lookup using this dot notation
+     * This selected fields implicitly declare the output fieldNames for the bolt based.
+     * @param commaSeparatedKeys
+     * @return
+     */
+    public RealtimeJoinBolt select(String commaSeparatedKeys) {
+        String[] fieldNames = commaSeparatedKeys.split(",");
+
+        outputFields = new FieldSelector[fieldNames.length];
+        for (int i = 0; i < fieldNames.length; i++) {
+            outputFields[i] = new FieldSelector(fieldNames[i]);
+        }
+        return this;
+    }
+
+    /** Convenience method for Streamline that prefixes each keyname with 'streamline-event.'
+     *
+     * @param commaSeparatedKeys
+     * @return
+     */
+    public RealtimeJoinBolt streamlineSelect(String commaSeparatedKeys) {
+        String prefixedKeys = convertToStreamLineKeys(commaSeparatedKeys);
+        streamLineProjection = true;
+        return  select(prefixedKeys);
+    }
+
+    public RealtimeJoinBolt withOutputStream(String streamName) {
+        this.outputStream = streamName;
+        return this;
+    }
+
+    @Override
+    public void execute(Tuple tuple) {
+        if (timeBasedRetention) {
+            expireAndAckTimedOutEntries(lookupBuffer);
```

Comment:
Incase of time based retention, expiring timedout entries occur only when next tuple comes, else all the expired tuples lie in the buffer.